File: src/prototype/web/tests/arm_extracted.test.ts
// Tests for arm_extracted flag — docs/design/state_machines_v1.md §4.1.

import { describe, expect, it } from "vitest";
import {
  DEFAULT_ARM_EXTRACTED_CONFIG,
  INITIAL_ARM_EXTRACTED,
  updateArmExtracted,
  type ArmExtractedInputs,
  type ArmExtractedState,
} from "../src/state/arm_extracted.js";
import type { HandFSM } from "../src/state/hand_fsm.js";
import type { GripZone } from "../src/input/intent.js";

function grippedAt(zone: GripZone, side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "GRIPPED" as const,
    target: zone,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}

function idleHand(side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "IDLE" as const,
    target: null,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}

function inputs(over: Partial<ArmExtractedInputs>): ArmExtractedInputs {
  return {
    nowMs: 0,
    dtMs: 16.67,
    bottomLeftHand: idleHand("L"),
    bottomRightHand: idleHand("R"),
    triggerL: 0,
    triggerR: 0,
    attackerHip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0 },
    defenderBaseHold: false,
    ...over,
  };
}

describe("arm_extracted transitions (§4.1)", () => {
  it("grip on SLEEVE_R with strong pull for 1.5s → right arm extracted", () => {
    let s: ArmExtractedState = INITIAL_ARM_EXTRACTED;
    const frames = Math.ceil(DEFAULT_ARM_EXTRACTED_CONFIG.requiredSustainMs / 16.67);
    for (let i = 0; i < frames; i += 1) {
      s = updateArmExtracted(s, inputs({
        nowMs: i * 16.67,
        bottomLeftHand: grippedAt("SLEEVE_R", "L"),
        triggerL: 0.8,
        attackerHip: { hip_angle_target: 0, hip_push: -0.5, hip_lateral: 0 },
      }));
    }
    expect(s.right).toBe(true);
    expect(s.left).toBe(false);
  });

  it("weak trigger fails to extract even with sustained time", () => {
    let s: ArmExtractedState = INITIAL_ARM_EXTRACTED;
    for (let i = 0; i < 120; i += 1) {
      s = updateArmExtracted(s, inputs({
        nowMs: i * 16.67,
        bottomLeftHand: grippedAt("SLEEVE_R", "L"),
        triggerL: 0.3,
        attackerHip: { hip_angle_target: 0, hip_push: -0.5, hip_lateral: 0 },
      }));
    }
    expect(s.right).toBe(false);
  });

  it("no hip pull (hip_push ≥ 0) fails to sustain", () => {
    let s: ArmExtractedState = INITIAL_ARM_EXTRACTED;
    for (let i = 0; i < 120; i += 1) {
      s = updateArmExtracted(s, inputs({
        nowMs: i * 16.67,
        bottomLeftHand: grippedAt("SLEEVE_R", "L"),
        triggerL: 0.8,
        attackerHip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0 },
      }));
    }
    expect(s.right).toBe(false);
  });

  it("releasing the grip clears a previously-set flag", () => {
    // Bring right arm to extracted first.
    let s: ArmExtractedState = INITIAL_ARM_EXTRACTED;
    const buildFrames = Math.ceil(DEFAULT_ARM_EXTRACTED_CONFIG.requiredSustainMs / 16.67);
    for (let i = 0; i < buildFrames; i += 1) {
      s = updateArmExtracted(s, inputs({
        nowMs: i * 16.67,
        bottomLeftHand: grippedAt("SLEEVE_R", "L"),
        triggerL: 0.9,
        attackerHip: { hip_angle_target: 0, hip_push: -0.6, hip_lateral: 0 },
      }));
    }
    expect(s.right).toBe(true);

    // Now release (hand IDLE).
    s = updateArmExtracted(s, inputs({
      nowMs: buildFrames * 16.67 + 16.67,
      bottomLeftHand: idleHand("L"),
    }));
    expect(s.right).toBe(false);
  });

  it("5s auto-reset clears the flag (briefly) even if the pull continues", () => {
    // Observe the exact tick after 5s: the flag must flip from true → false.
    // The attacker can of course re-set it later, but the auto-reset itself
    // is the contract we're testing.
    let s: ArmExtractedState = INITIAL_ARM_EXTRACTED;
    const step = 50;

    // Build up to true.
    const buildUpFrames = Math.ceil(DEFAULT_ARM_EXTRACTED_CONFIG.requiredSustainMs / step);
    for (let i = 0; i < buildUpFrames; i += 1) {
      s = updateArmExtracted(s, inputs({
        nowMs: i * step,
        dtMs: step,
        bottomLeftHand: grippedAt("SLEEVE_R", "L"),
        triggerL: 0.9,
        attackerHip: { hip_angle_target: 0, hip_push: -0.6, hip_lateral: 0 },
      }));
    }
    expect(s.right).toBe(true);
    const setAt = s.rightSetAtMs;

    // Advance to just past the auto-reset boundary.
    const resetTick = setAt + DEFAULT_ARM_EXTRACTED_CONFIG.autoResetAfterMs;
    s = updateArmExtracted(s, inputs({
      nowMs: resetTick,
      dtMs: step,
      bottomLeftHand: grippedAt("SLEEVE_R", "L"),
      triggerL: 0.9,
      attackerHip: { hip_angle_target: 0, hip_push: -0.6, hip_lateral: 0 },
    }));
    expect(s.right).toBe(false);
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/ArmExtractedTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/unit/arm_extracted.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite.
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using NUnit.Framework;

namespace BJJSimulator.Tests
{
    [TestFixture]
    public class ArmExtractedTest
    {
        static HandFSM GrippedAt(GripZone zone, HandSide side) => new HandFSM
        {
            Side             = side,
            State            = HandState.Gripped,
            Target           = zone,
            StateEnteredMs   = 0,
            ReachDurationMs  = 0,
            LastParriedZone  = GripZone.None,
            LastParriedAtMs  = BJJConst.SentinelTimeMs,
        };

        static HandFSM IdleHand(HandSide side) => new HandFSM
        {
            Side             = side,
            State            = HandState.Idle,
            Target           = GripZone.None,
            StateEnteredMs   = 0,
            ReachDurationMs  = 0,
            LastParriedZone  = GripZone.None,
            LastParriedAtMs  = BJJConst.SentinelTimeMs,
        };

        static ArmExtractedInputs BaseInputs(
            long nowMs          = 0,
            float dtMs          = 16.67f,
            HandFSM? leftHand   = null,
            HandFSM? rightHand  = null,
            float triggerL      = 0f,
            float triggerR      = 0f,
            float hipPush       = 0f,
            bool  defenderBase  = false) =>
            new ArmExtractedInputs
            {
                NowMs            = nowMs,
                DtMs             = dtMs,
                BottomLeftHand   = leftHand  ?? IdleHand(HandSide.L),
                BottomRightHand  = rightHand ?? IdleHand(HandSide.R),
                TriggerL         = triggerL,
                TriggerR         = triggerR,
                AttackerHip      = new HipIntent { HipPush = hipPush },
                DefenderBaseHold = defenderBase,
            };

        // Helper: drive until flag is true via the standard extract sequence.
        ArmExtractedState BuildExtracted(float step, out long setAt)
        {
            var s = ArmExtractedState.Initial;
            int frames = (int)System.Math.Ceiling(ArmExtractedConfig.Default.RequiredSustainMs / step);
            long now   = 0;
            for (int i = 0; i < frames; i++, now = (long)(i * step))
            {
                s = ArmExtractedOps.Update(s, BaseInputs(
                    nowMs:    now,
                    dtMs:     step,
                    leftHand: GrippedAt(GripZone.SleeveR, HandSide.L),
                    triggerL: 0.9f,
                    hipPush: -0.6f));
            }
            setAt = s.LeftSetAtMs;
            return s;
        }

        // -------------------------------------------------------------------------

        [Test]
        public void GripOnSleeveR_WithStrongPull_For1500ms_Sets_RightArmExtracted()
        {
            float step  = 16.67f;
            int   frames = (int)System.Math.Ceiling(ArmExtractedConfig.Default.RequiredSustainMs / step);
            var   s     = ArmExtractedState.Initial;
            for (int i = 0; i < frames; i++)
            {
                s = ArmExtractedOps.Update(s, BaseInputs(
                    nowMs:    (long)(i * step),
                    dtMs:     step,
                    leftHand: GrippedAt(GripZone.SleeveR, HandSide.L),
                    triggerL: 0.8f,
                    hipPush: -0.5f));
            }
            Assert.IsTrue(s.Right, "right should be extracted");
            Assert.IsFalse(s.Left, "left should not be extracted");
        }

        [Test]
        public void WeakTrigger_FailsToExtract_EvenWithSustainedTime()
        {
            var s = ArmExtractedState.Initial;
            for (int i = 0; i < 120; i++)
            {
                s = ArmExtractedOps.Update(s, BaseInputs(
                    nowMs:    (long)(i * 16.67f),
                    leftHand: GrippedAt(GripZone.SleeveR, HandSide.L),
                    triggerL: 0.3f,
                    hipPush: -0.5f));
            }
            Assert.IsFalse(s.Right);
        }

        [Test]
        public void NoHipPull_FailsToSustain()
        {
            var s = ArmExtractedState.Initial;
            for (int i = 0; i < 120; i++)
            {
                s = ArmExtractedOps.Update(s, BaseInputs(
                    nowMs:    (long)(i * 16.67f),
                    leftHand: GrippedAt(GripZone.SleeveR, HandSide.L),
                    triggerL: 0.8f,
                    hipPush: 0f));
            }
            Assert.IsFalse(s.Right);
        }

        [Test]
        public void ReleasingGrip_Clears_PreviouslySetFlag()
        {
            long setAt;
            var s = BuildExtracted(16.67f, out setAt);
            Assert.IsTrue(s.Right, "precondition: right was extracted");

            s = ArmExtractedOps.Update(s, BaseInputs(
                nowMs: (long)(s.RightSustainMs + 16.67f),
                leftHand: IdleHand(HandSide.L)));
            Assert.IsFalse(s.Right, "should be cleared after releasing grip");
        }

        [Test]
        public void AutoReset_Clears_FlagAfter5s_EvenWhilePulling()
        {
            float step = 50f;
            long  setAt;
            var   s = BuildExtracted(step, out setAt);
            Assert.IsTrue(s.Right, "precondition: flag is set");
            Assert.AreNotEqual(BJJConst.SentinelTimeMs, s.RightSetAtMs, "setAtMs must not be sentinel");

            long resetTick = s.RightSetAtMs + ArmExtractedConfig.Default.AutoResetAfterMs;
            s = ArmExtractedOps.Update(s, BaseInputs(
                nowMs:    resetTick,
                dtMs:     step,
                leftHand: GrippedAt(GripZone.SleeveR, HandSide.L),
                triggerL: 0.9f,
                hipPush: -0.6f));
            Assert.IsFalse(s.Right, "auto-reset should have cleared the flag");
        }
    }
}
